### PR TITLE
chore: watermark and eventtime in windowing operations

### PR DESCRIPTION
**Pull request by ashwinidulams**
for window operations today, we use window start time for both eventtime and watermark. with this PR, that will be changed to (window.endtime - millisecond)

### CODE REVIEW

**Review by vigith — APPROVED** [review]
Could you please update the reduce doc too on how we set the end-time of reduced outputs and how to interpret it?

**Comment by ashwinidulams** [comment]
> 

sure will do.

File: docs/user-guide/user-defined-functions/reduce/reduce.md [hunk 1]
```diff
@@ -68,6 +68,22 @@ It is wrong to give a `parallelism` > `1` if it is a _non-keyed_ vertex (`keyed:
 There are a couple of [examples](examples.md) that demonstrates Fixed windows, Sliding windows,
 chaining of windows, keyed streams, etc.
 
+## time characteristics
```
Comment (vigith) [inline docs/user-guide/user-defined-functions/reduce/reduce.md:71]:
```suggestion
## Time Characteristics
```

File: docs/user-guide/user-defined-functions/reduce/reduce.md [hunk 2]
```diff
@@ -68,6 +68,22 @@ It is wrong to give a `parallelism` > `1` if it is a _non-keyed_ vertex (`keyed:
...
+
+All windowing operations generate new records as an output of reduce operations. event time and watermark
```
Comment (vigith) [inline docs/user-guide/user-defined-functions/reduce/reduce.md:73]:
```suggestion
All windowing operations generate new records as an output of reduce operations. Event-time and Watermark 
```

File: docs/user-guide/user-defined-functions/reduce/reduce.md [hunk 3]
```diff
@@ -68,6 +68,22 @@ It is wrong to give a `parallelism` > `1` if it is a _non-keyed_ vertex (`keyed:
 There are a couple of [examples](examples.md) that demonstrates Fixed windows, Sliding windows,
 chaining of windows, keyed streams, etc.
 
+## time characteristics
+
+All windowing operations generate new records as an output of reduce operations. event time and watermark 
+are two main primitives that determine how the time propagates in a streaming application. so for all new 
+records generated in a reduce operation, event time is set to the end time of the window. 
+
+for example, for a reduce operation over a keyed/non-keyed window with start and end defined by
```
Comment (vigith) [inline docs/user-guide/user-defined-functions/reduce/reduce.md:77]:
```suggestion
For example, for a reduce operation over a keyed/non-keyed window with a start and end defined by 
```

File: docs/user-guide/user-defined-functions/reduce/reduce.md [hunk 4]
```diff
@@ -68,6 +68,22 @@ It is wrong to give a `parallelism` > `1` if it is a _non-keyed_ vertex (`keyed:
...
+All windowing operations generate new records as an output of reduce operations. event time and watermark 
+are two main primitives that determine how the time propagates in a streaming application. so for all new 
+records generated in a reduce operation, event time is set to the end time of the window. 
+
+for example, for a reduce operation over a keyed/non-keyed window with start and end defined by 
+`[2031-09-29T18:47:00Z, 2031-09-29T18:48:00Z)`, event time for all the records generated will be set to 
+`2031-09-29T18:47:59.999Z` since millisecond is the smallest granularity (as of now) event time is set to 
+the last timestamp that belongs to a window. 
+
+watermark is treated similarly. watermark is set to the last timestamp for a given window.
```
Comment (vigith) [inline docs/user-guide/user-defined-functions/reduce/reduce.md:82]:
```suggestion
Watermark is treated similarly, the watermark is set to the last timestamp for a given window. 
```

File: docs/user-guide/user-defined-functions/reduce/reduce.md [hunk 5]
```diff
@@ -68,6 +68,22 @@ It is wrong to give a `parallelism` > `1` if it is a _non-keyed_ vertex (`keyed:
...
+are two main primitives that determine how the time propagates in a streaming application. so for all new 
+records generated in a reduce operation, event time is set to the end time of the window. 
+
+for example, for a reduce operation over a keyed/non-keyed window with start and end defined by 
+`[2031-09-29T18:47:00Z, 2031-09-29T18:48:00Z)`, event time for all the records generated will be set to 
+`2031-09-29T18:47:59.999Z` since millisecond is the smallest granularity (as of now) event time is set to 
+the last timestamp that belongs to a window. 
+
+watermark is treated similarly. watermark is set to the last timestamp for a given window. 
+so for the example above, the value of watermark will be set to the last timestamp i.e. `2031-09-29T18:47:59.999Z`
```
Comment (vigith) [inline docs/user-guide/user-defined-functions/reduce/reduce.md:83]:
```suggestion
So for the example above, the value of the watermark will be set to the last timestamp, i.e., `2031-09-29T18:47:59.999Z`.
```

**Review by vigith — APPROVED** [review]

**Pull request merged**